bot/exts/info/doc/_parsing.py

```diff
@@ -0,0 +1,328 @@
+from __future__ import annotations
+
+import logging
+import re
+import string
+import textwrap
+from functools import partial
+from typing import Callable, Collection, Container, Iterable, List, Optional, TYPE_CHECKING, Union
+
+from bs4 import BeautifulSoup
+from bs4.element import NavigableString, PageElement, Tag
+
+from bot.utils.helpers import find_nth_occurrence
+from ._html import Strainer
+from ._markdown import DocMarkdownConverter
+if TYPE_CHECKING:
+    from ._cog import DocItem
+
+log = logging.getLogger(__name__)
+
+_MAX_SIGNATURE_AMOUNT = 3
+
+_UNWANTED_SIGNATURE_SYMBOLS_RE = re.compile(r"\[source]|\\\\|¶")
+_WHITESPACE_AFTER_NEWLINES_RE = re.compile(r"(?<=\n\n)(\s+)")
+_PARAMETERS_RE = re.compile(r"\((.+)\)")
+
+_SEARCH_END_TAG_ATTRS = (
+    "data",
+    "function",
+    "class",
+    "exception",
+    "seealso",
+    "section",
+    "rubric",
+    "sphinxsidebar",
+)
+
+_NO_SIGNATURE_GROUPS = {
+    "attribute",
+    "envvar",
+    "setting",
+    "tempaltefilter",
+    "templatetag",
+    "term",
+}
+_EMBED_CODE_BLOCK_LINE_LENGTH = 61
+# _MAX_SIGNATURE_AMOUNT code block wrapped lines with py syntax highlight
+_MAX_SIGNATURES_LENGTH = (_EMBED_CODE_BLOCK_LINE_LENGTH + 8) * _MAX_SIGNATURE_AMOUNT
+# Maximum discord message length - signatures on top
+_MAX_DESCRIPTION_LENGTH = 2000 - _MAX_SIGNATURES_LENGTH
+_TRUNCATE_STRIP_CHARACTERS = "!?:;." + string.whitespace
+_BRACKET_PAIRS = {
+    "{": "}",
+    "(": ")",
+    "[": "]",
+}
+
+
+def _split_parameters(parameters_string: str) -> List[str]:
```

Comment:
Signatures aren't always valid syntax because of optionals etc.

Reply:
It may be  more likely for the current parser to split incorrectly than it is for a signature to be invalid syntax. I wonder if it's worth the complexity to use ast and fall back to the current parser. I haven't seen how this performs with practical data.

Reply:
An another problem with ast is that it destroy the information of how the initial source looked, but the custom parser just lets it through.
I don't think there are many cases where the current parser will fail after the string errors are handled, it won't be able to split conditionals because it can't tell apart an optional bracket or a list arg but should do fine in other cases as it only needs to keep count of the brackets

Reply:
That's a good point. I'm still not particularly confident in this split function but we can roll with it and fix parsing bugs along the way if we encounter any common cases.